### PR TITLE
Correct descriptions of CAMLreturn and CAMLreturn0 in interfacing C page, Fixes #7179

### LINKED ISSUE #7179: Interfacing with C missing documentation and helper macros and functions
**Original bug ID:** 7179
**Reporter:** goswin
**Status:** acknowledged (set by @damiendoligez on 2016-03-14T13:34:46Z)
**Resolution:** open
**Priority:** normal
**Severity:** minor
**Category:** documentation
**Monitored by:** @jmeber

## Bug description

When an ocaml function is used as callback from C code and the last thing the callback does is call a C stub returning unit and the return register happens to contain ((x & 3) == 2) then this is taken as an exception. When caml_callback() is used this carshes and when caml_callback_exn is garbage is returned from Extract_exception().

Ocaml should sanitize the return register after calling the C stub.

## Steps to reproduce

% make
ocamlopt -g -c -o main.cmx main.ml
gcc -O2 -W -Wall -g -c -o stubs.o stubs.c
ocamlopt -o test main.cmx stubs.o
./test
test(0x61ea78)
fail
test: callback got exception 0x0

==> Makefile <==
        ocamlopt -o $@ $+

%.cmx: %.ml
        ocamlopt -g -c -o $@ $<

%.o: %.c
        gcc -O2 -W -Wall -g -c -o $@ $<

clean:
        rm -f test *.o *.cmx *.cmi *~

==> main.ml <==
external fail : unit -> unit = "fail"
external test : (unit -> unit) -> unit = "test"

let callback () =
  fail ()
   
let () =
  test callback

==> stubs.c <==
    CAMLparam1(callback);
    CAMLlocal1(res);
    printf("%s(0x%lx)\n", __PRETTY_FUNCTION__, callback);
    res = caml_callback_exn(callback, Val_unit);
    if (Is_exception_result(res)) {
        res = Extract_exception(res);
        fprintf(stderr, "%s: callback got exception 0x%lx\n", __PRETTY_FUNCTION__, res);
    }
    CAMLreturn0;
}

## Additional information

https://github.com/mrvn/ocam-problems/tree/false-exception

### PR DESCRIPTION
**Pull request by kirisky**
This PR fixes #7179.

Correct the description of CAMLreturn and CAMLreturn0 in the Interfacing C page and memory.h file.

### CODE REVIEW

**Comment by gadmm** [comment]
Thanks for looking into this issue. I do not think that the correct change is to advise to use `CAMLreturn(Val_unit)` instead of `CAMLreturn0`. `CAMLreturn0` is the correct macro to end functions that return `void`.

Looking at the linked issue, I think the solution is to warn somewhere that using a function that returns `void` as an `extern` function in OCaml is a type error—C functions that return the OCaml type `unit` should return the C type `value`.

**Comment by kirisky** [comment]
> Thanks for looking into this issue. I do not think that the correct change is to advise to use `CAMLreturn(Val_unit)` instead of `CAMLreturn0`. `CAMLreturn0` is the correct macro to end functions that return `void`.
> 
> Looking at the linked issue, I think the solution is to warn somewhere that using a function that returns `void` as an `extern` function in OCaml is a type error—C functions that return the OCaml type `unit` should return the C type `value`.

Thank you very much to point this out, @gadmm! 😃 
I put the warning description in a new place and added an example for it. Could you help to take a look?

**Comment by OlivierNicole** [comment]
I have been bitten in the past when I tried to use a void-returning function from OCaml, so I confirm that such an addition to the manual would be useful.

File: manual/src/cmds/intf-c.etex
Comment (OlivierNicole) [inline manual/src/cmds/intf-c.etex]: This should not be changed. It illustrates the rule that function having parameters or local variables of type `value` should use the macros `CAMLreturn*` and `CAMLparam*`. Such functions are allowed to have return type `void`, so long as they are not called from OCaml. Note that the change would be a C type error, since we are returning a `value` from a `void` function.

File: manual/src/cmds/intf-c.etex
Comment (OlivierNicole) [inline manual/src/cmds/intf-c.etex]: This is in section “Callbacks from C to OCaml”. I feel that this is not quite the right place, as we want to warn specifically about C functions called from OCaml.

I would have put this as a warning (`\paragraph{Warning:} ...`) shortly after the introduction of the `CAMLreturn*` and `CAMLparam*` macros, in section 5.1, and phrased it similarly to what @gasche proposed in the original issue:

> clarify the documentation of the C interface in the manual to explain that CAMLreturn0 should only be used for internal FFI procedures of type void, and that CAMLreturn(Val_unit) should be used for functions that return an OCaml unit value.

Not that the obligation of returning `value`s to OCaml is already specified near the beginning of the chapter, in section 1.2:

> User primitives with arity n ≤ 5 are implemented by C functions that take n arguments of type value, and return a result of type value.

However, I am in favor of reminding it for the specific case of `CAMLreturn0` vs unit, when this macro is introduced, as the mistake is particularly easy to make.

**Review by OlivierNicole — CHANGES_REQUESTED** [review]

File: manual/src/cmds/intf-c.etex
Comment (gadmm) [reply manual/src/cmds/intf-c.etex]: Yes, this sounds like the right thing to do. The original issue mentioned `caml_callback_exn` but this is a red herring, other things can probably go wrong by having garbage as a return value, such as crashing the GC.

Comment (OlivierNicole) [reply manual/src/cmds/intf-c.etex]: Yes, that is exactly what I experienced.

Comment (kirisky) [reply manual/src/cmds/intf-c.etex]: Updated. Please help to review @gadmm and @OlivierNicole
Thanks, guys! 😄

File: manual/src/cmds/intf-c.etex
Comment (OlivierNicole) [inline manual/src/cmds/intf-c.etex]:
```suggestion
\paragraph{Warning:} "CAMLreturn0" should only be used for internal procedures that return void.
"CAMLreturn(Val_unit)" should be used for functions that return an OCaml unit value. Primitives (C functions that can be called from OCaml) should never return void.
```
Suggestions: simplify the structure of these two sentences. I also propose to avoid the use of the word “FFI” in favor of “primitive” for consistency with the rest of the chapter. My suggestion also contains an additional sentence to insist on the fact that functions called from OCaml should never return void.

**Review by OlivierNicole — COMMENTED** [review]
Thank you for the update. I suggest a few simplifications to make the paragraph even more clear. Let me know what you think.

File: manual/src/cmds/intf-c.etex
Comment (kirisky) [reply manual/src/cmds/intf-c.etex]: Thank you very much, @OlivierNicole! 😄 
Updated https://github.com/ocaml/ocaml/pull/11894/commits/78b817d3123f27abd95ebf2ca198576bc508fddb
Please help to review!

**Review by gadmm — APPROVED** [review]
This addition looks good to me (cc @gasche)

**Review by gasche — APPROVED** [review]
Approved on behalf of @gadmm. Thanks!

**Comment by gadmm** [comment]
It was missing the correct change entry and perhaps some squashing, though.

**Comment by gasche** [comment]
I squashed the commits through the Github UI, and the Changes entry looked okay to me.
(Note: in absence of any other indication I interpret approval as "this can be merged now"; if there are conditions to fulfill before merging it's helpful to point them out explicitly on approval.)

**Pull request merged**
